docs/src/nondiff_points.md

```diff
@@ -0,0 +1,149 @@
+For a function like `abs` or `x -> x<0 ? x : -2x` setting the non-differentiable point to either side would result in it leaving that maxima.
+
+Further, a nice (albeit generally intractable) algorithm for finding the global optima is to take the set of all stationary points (i.e. points where the derivative is zero), combine that with the boundary points and evaluate the primal function at all of them.
+The extrema of this set are the global optima.
+This algorithm is only guaranteed correct for functions that are differentiable everywhere *or* that apply the sub/super-derivative convention and make all non-differentiable local optima claim to have a zero derivative.
+It is also correct if you make other non-differentiable points have zero derivative, just slower.
+
+### It is sometimes the non-zero, especially if it boarders a flat section
+
+If a function has derivative zero in some large section of it's domain, like `relu` or `x->clamp(x, -1, 1)`,
+a case can be made for choosing the non-zero derivative branch.
+Depending exactly on the larger function being optimized, it is often the case that a zero gradient for this function means a total zero gradient (e.g. if the greater function is a chain of composed calls).
+So once things move into the flat-region they stay there.
+If we chose the other branch we move them into (just barely) the nonflat region.
+If moving into the non-flat region was good they will move further there later.
+If it is not good then they well be rapidly cast deeper into the flat region and we will not be at this boundary case.
+
+### It is sometimes the mean
+Either the mean of the branches, or the overall mean of the sub/super-differential (which may be equal).
+
+A nice advantage of it is that it will agree with the result you will get from central finite differencing.
+Which is what [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl) defaults to.
+
+### All else being equal it is just one of the branches
+Pick one.
+It will be fast.
+It will agree with either the forwards or reverse finite differencing results.
+
+### It is almost never `Inf`, `NaN` or an error
+Practically speaking, people are generally using AD to perform some gradient descent like optimization procedure.
+`NaN` and `Inf` do not generally take us to nice places.
+Erroring can be worse, especially if it is a local minima -- the optimization fully converges to that minima and then throws an error rather than reporting the result.
+If the primal function takes a non-finite value or errors on one side, then we are in the case that we are at a boundry of the domain.
+Which means we are free to chose *any* value.
+In particular we often want to chose value of the derivative from **the other side where the primal is defined.**
+
+
+## Why is the sub/super-differential convention permissible?
+
+Math isn't real, and AD doesn't do calculus.
+We are trying to accomplish some goal, and this is approach works well.
```

Comment:
```suggestion
Math isn't real, and AD doesn't do calculus.
We are trying to accomplish some goal, and this approach works well. One way to think about this 
convention is to take an infinitely sharp discontinuity in a function and replace that discontinuty with
a very (infinitesimally) small, smooth corner.  Due to the intermediate value theorem, we can then 
say that over this tiny interval, any derivative between the two extremes is realized and we are free 
to pick any one of them that we find useful as the 'canonical' value.
```

Reply:
Maybe it'd be good to include a gif here:
![abs](https://user-images.githubusercontent.com/29157027/127267988-92ca0899-ca33-4dd6-af45-e5a085283c83.gif)

```julia
julia> using Plots

julia> let xs = range(-5, 5, length=10_000)
           hyp(x; a, b=a) = √(a^2 + (a*x/b)^2)
           @gif for a ∈ range(1.0, 0.001, length=150)
               plot(xs, hyp.(xs; a), ticks=false, legend=false)
           end 
       end
```